Index: cloud-init-0.7.9/systemd/cloud-init.service
===================================================================
--- cloud-init-0.7.9.orig/systemd/cloud-init.service
+++ cloud-init-0.7.9/systemd/cloud-init.service
@@ -4,7 +4,8 @@ Wants=cloud-init-local.service
 Wants=sshd-keygen.service
 Wants=sshd.service
 After=cloud-init-local.service
-After=networking.service
+After=network.service
+After=NetworkManager.service
 Before=network-online.target
 Before=sshd-keygen.service
 Before=sshd.service
