Index: cloud-init-0.7.8/tests/unittests/test_handler/test_handler_apt_source_v3.py
===================================================================
--- cloud-init-0.7.8.orig/tests/unittests/test_handler/test_handler_apt_source_v3.py
+++ cloud-init-0.7.8/tests/unittests/test_handler/test_handler_apt_source_v3.py
@@ -947,6 +947,9 @@ deb http://ubuntu.com/ubuntu/ xenial-pro
         self.assertEqual(
             orig, cc_apt_configure.disable_suites(["proposed"], orig, rel))
 
+    # https://bugs.launchpad.net/cloud-init/+bug/1629149
+    import unittest
+    @unittest.skip('module not used on Fedora; test is env-sensitive')
     def test_apt_v3_mirror_search_dns(self):
         """test_apt_v3_mirror_search_dns - Test searching dns patterns"""
         pmir = "phit"
